zokrates_core/src/static_analysis/zir_propagation.rs

```diff
@@ -0,0 +1,1414 @@
+
+#[derive(Default)]
+pub struct ZirPropagator<'ast, T> {
+    constants: Constants<'ast, T>,
+}
+
+impl<'ast, T: Field> ZirPropagator<'ast, T> {
+    pub fn new() -> Self {
+        ZirPropagator::default()
+    }
+    pub fn propagate(p: ZirProgram<T>) -> ZirProgram<T> {
+        ZirPropagator::new().fold_program(p)
+    }
+}
+
+impl<'ast, T: Field> Folder<'ast, T> for ZirPropagator<'ast, T> {
+    fn fold_statement(&mut self, s: ZirStatement<'ast, T>) -> Vec<ZirStatement<'ast, T>> {
+        match s {
+            ZirStatement::Assertion(e) => match self.fold_boolean_expression(e) {
+                BooleanExpression::Value(true) => vec![],
+                e => vec![ZirStatement::Assertion(e)],
+            },
+            ZirStatement::Definition(a, e) => {
+                let e = self.fold_expression(e);
+                match e {
+                    ZirExpression::FieldElement(FieldElementExpression::Number(..))
+                    | ZirExpression::Boolean(BooleanExpression::Value(..))
+                    | ZirExpression::Uint(UExpression {
+                        inner: UExpressionInner::Value(..),
+                        ..
+                    }) => {
+                        self.constants.insert(a.id, e);
+                        vec![]
+                    }
+                    _ => {
+                        self.constants.remove(&a.id);
+                        vec![ZirStatement::Definition(a, e)]
+                    }
+                }
+            }
+            ZirStatement::MultipleDefinition(assignees, list) => {
+                for a in &assignees {
+                    self.constants.remove(&a.id);
+                }
+                vec![ZirStatement::MultipleDefinition(
+                    assignees,
+                    self.fold_expression_list(list),
+                )]
+            }
+            _ => fold_statement(self, s),
+        }
+    }
+
+    fn fold_field_expression(
+        &mut self,
+        e: FieldElementExpression<'ast, T>,
+    ) -> FieldElementExpression<'ast, T> {
+        match e {
+            FieldElementExpression::Number(n) => FieldElementExpression::Number(n),
+            FieldElementExpression::Identifier(id) => match self.constants.get(&id) {
+                Some(ZirExpression::FieldElement(FieldElementExpression::Number(v))) => {
+                    FieldElementExpression::Number(v.clone())
+                }
+                _ => FieldElementExpression::Identifier(id),
+            },
+            FieldElementExpression::Select(e, box index) => {
+                let index = self.fold_uint_expression(index);
+                let e: Vec<FieldElementExpression<'ast, T>> = e
+                    .into_iter()
+                    .map(|e| self.fold_field_expression(e))
+                    .collect();
+
+                match index.into_inner() {
+                    UExpressionInner::Value(v) => {
+                        e.get(v as usize).cloned().expect("index out of bounds")
```

Comment:
Is this panic reachable? I suspect it is, if the index is something like `if [a, 1] != [b, 0] then 1000000000 else 0`. That would suggest this propagation step has to be failible, which would require a `result_folder` for ZIR. Or we pretend we didn't see and return the original, which will fail at compile time, but that sounds less nice.

Reply:
Yeah, you're right. I think we should use a `result_folder` here